src/mo_types/cons.ml

```diff
@@ -17,6 +17,12 @@ module Stamps = Env.Make(String)
 
 let stamps : int Stamps.t ref = ref Stamps.empty
 
+let session f =
+  let original = !stamps in
+  let result = f () in
+  stamps := original;
+  result
```

Comment:
```suggestion
let session f =
  let original = !stamps in
  try let result = f () in
       stamps := original;
       result
  match e -> begin
     stamps := original;
     raise e
  end
```
Looks good but probably worth protecting the exception exit too, with something like the above.